{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d1305c1",
   "metadata": {},
   "source": [
    "# Differential Gene Regulatory Network Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1222392",
   "metadata": {},
   "source": [
    "The first step is to import the required packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a893b2d9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"recount\")\n",
    "BiocManager::install(\"recount3\")\n",
    "BiocManager::install(\"limma\")\n",
    "BiocManager::install(\"DESeq2\")\n",
    "BiocManager::install(\"vsn\")\n",
    "BiocManager::install(\"biomaRt\")\n",
    "BiocManager::install(\"fgsea\")\n",
    "BiocManager::install(\"edgeR\")\n",
    "\n",
    "install.packages(\"visNetwork\")\n",
    "install.packages(\"ggplot2\")\n",
    "install.packages(\"dplyr\")\n",
    "install.packages(\"ggrepel\")\n",
    "install.packages(\"gridExtra\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21af9f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "library('recount')    # To process gene expression\n",
    "library('recount3')   # To process gene expression\n",
    "library('limma')      # To average replicates\n",
    "library('DESeq2')     # To do the mean-variance correction\n",
    "library('vsn')        # To visualize mean-variance plots\n",
    "library('biomaRt')    # To convert ensembl gene/protein ids to hugo (hgnc) symbols\n",
    "library('visNetwork') # For network visualization \n",
    "library('fgsea')      # For gene set enrichment analysis \n",
    "library('ggplot2')    # To build the bubble plot for gene enrichment\n",
    "library('dplyr')      # For data manipulation\n",
    "library('edgeR')\n",
    "library('ggrepel')    # Prevents overlapping text\n",
    "library('gridExtra')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ab58d6f",
   "metadata": {},
   "source": [
    "The second step is to import TCGA Human Pancreatic Adenocarcinoma and GTEx Pancreas data from Recount 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d6cef75",
   "metadata": {},
   "outputs": [],
   "source": [
    "rse_geneTumorDataOriginal <- recount3::create_rse_manual( project = \"PAAD\",\n",
    "  project_home = \"data_sources/tcga\",\n",
    "  organism = \"human\",\n",
    "  annotation = \"gencode_v26\",\n",
    "  type = \"gene\"\n",
    ")\n",
    "\n",
    "rse_geneNormalDataOriginal <- recount3::create_rse_manual(\n",
    "  project = \"PANCREAS\",\n",
    "  project_home = \"data_sources/gtex\",\n",
    "  organism = \"human\",\n",
    "  annotation = \"gencode_v26\",\n",
    "  type = \"gene\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1169f19",
   "metadata": {},
   "source": [
    "The next step is to import a list of Human TFs and lists of X chromosome genes and Y chromosome genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "708f2b04",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_tf <- as.data.frame(read.table(\"http://humantfs.ccbr.utoronto.ca/download/v_1.01/TF_names_v_1.01.txt\", header = FALSE, sep = \"\\n\", dec = \".\"))\n",
    "human_tf=human_tf[[\"V1\"]]\n",
    "\n",
    "Ygene <- as.vector(read.table(\"https://www.genenames.org/cgi-bin/download/custom?col=gd_app_sym&status=Approved&chr=Y&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&where=(gd_pub_chrom_map%20not%20like%20%27%25patch%25%27%20and%20gd_pub_chrom_map%20not%20like%20%27%25alternate%20reference%20locus%25%27)&submit=submit\", header = TRUE, sep = \"\\n\"))\n",
    "Xgene <- as.vector(read.table(\"https://www.genenames.org/cgi-bin/download/custom?col=gd_app_sym&status=Approved&chr=X&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&where=(gd_pub_chrom_map%20not%20like%20%27%25patch%25%27%20and%20gd_pub_chrom_map%20not%20like%20%27%25alternate%20reference%20locus%25%27)&submit=submit\", header = TRUE, sep = \"\\n\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eaf1240",
   "metadata": {},
   "source": [
    "The \"vsdcreator\" function takes in a Random Summarized Experiment object, subsets the genes so that only genes of the specified gene type remain, removed the genes that have less than or equal to one count across all samples, and outputs meanSd plots and count matrix if desired. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6920f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "vsdcreator <- function(rseobject, genetype, includePlots, includecountMat, isTumor){\n",
    "  rse_gene <- rseobject\n",
    "  rowDataRse=rowData(rse_gene)\n",
    "  colDataRse=colData(rse_gene)\n",
    "  \n",
    "  geneIds = c()\n",
    "  geneNames = c()\n",
    "  geneEntrezIds = c()\n",
    "  \n",
    "  for(i in 1:dim(rowDataRse)[1]){\n",
    "    if(rowDataRse@listData$gene_type[i][[1]][1] == genetype){\n",
    "      geneIds       = c(geneIds,i)\n",
    "      geneNames     = c(geneNames,rowDataRse@listData$gene_name[i][[1]][1])\n",
    "      geneEntrezIds = c(geneEntrezIds, substr(rowDataRse@listData$gene_id[i],1,15))\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  if(isTumor == TRUE){\n",
    "    tumorIds=c()\n",
    "    patientSymbol=c()\n",
    "    \n",
    "    for(i in 1:length(colDataRse@listData$tcga.gdc_cases.samples.sample_type)){\n",
    "      if(colDataRse@listData$tcga.gdc_cases.samples.sample_type[i] %in% c(\"Primary Tumor\",\"Metastatic\") && colDataRse@listData$tcga.gdc_cases.project.name[i] == \"Pancreatic Adenocarcinoma\"){\n",
    "        tumorIds = c(tumorIds,i)     \n",
    "        patientSymbol = c(patientSymbol, colDataRse@listData$tcga.gdc_cases.samples.portions.analytes.aliquots.submitter_id[i])\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    rse_gene=rse_gene[geneIds,tumorIds]\n",
    "  }\n",
    "  if(isTumor == FALSE){\n",
    "    rse_gene=rse_gene[geneIds,]\n",
    "  }\n",
    "  \n",
    "  \n",
    "  countMat=SummarizedExperiment::assay(rse_gene, 1)\n",
    "  \n",
    "  indFilter = which(rowSums(countMat) <= 1)\n",
    "  \n",
    "  rse_gene=rse_gene[-indFilter,]\n",
    "  \n",
    "  rowDataRse=rowData(rse_gene)\n",
    "  colDataRse=colData(rse_gene)\n",
    "  \n",
    "  countMat=SummarizedExperiment::assay(rse_gene, 1)\n",
    "  meanSdPlotBeforeCorrection <- meanSdPlot(countMat)\n",
    "  meanSdPlotBeforeCorrectionLogTrans <- meanSdPlot(log(countMat+1), ranks = FALSE)\n",
    "  \n",
    "  if (genetype == \"miRNA\"){\n",
    "    vsd <- varianceStabilizingTransformation(countMat, blind=FALSE)\n",
    "  } else{\n",
    "  vsd <- vst(countMat, blind=FALSE)\n",
    "  }\n",
    "  \n",
    "  meanSdPlot <- meanSdPlot(vsd)\n",
    "  \n",
    "  \n",
    "  if(isTumor == TRUE){\n",
    "    colnames(vsd)=patientSymbol\n",
    "  }\n",
    "  rownames(vsd)=geneNames[-indFilter]\n",
    "  \n",
    "  vsd=avearrays(t(vsd))\n",
    "  vsd=as.data.frame(t(vsd))\n",
    "  \n",
    "  if(isTumor == TRUE){\n",
    "    colnames(countMat)=patientSymbol\n",
    "  }\n",
    "  rownames(countMat)=geneNames[-indFilter]\n",
    "  countMat=avearrays(t(countMat))\n",
    "  countMat=as.data.frame(t(countMat))\n",
    "  \n",
    "  if (includePlots == FALSE & includecountMat == FALSE){\n",
    "    return(list(rse = rse_gene, vsdoutput = vsd))\n",
    "  }\n",
    "  if (includePlots == TRUE & includecountMat == FALSE){\n",
    "    return(list(rse = rse_gene,vsdoutput = vsd, meanSdPlotBeforeCorrectionoutput = meanSdPlotBeforeCorrection, meanSdPlotBeforeCorrectionLogTransoutput = meanSdPlotBeforeCorrectionLogTrans, meanSdPlotoutput = meanSdPlot))\n",
    "  }\n",
    "  \n",
    "  if (includePlots == TRUE & includecountMat == TRUE){\n",
    "    return(list(rse = rse_gene, vsdoutput = vsd, meanSdPlotBeforeCorrectionoutput = meanSdPlotBeforeCorrection, meanSdPlotBeforeCorrectionLogTransoutput = meanSdPlotBeforeCorrectionLogTrans, meanSdPlotoutput = meanSdPlot, countMatoutput = countMat))\n",
    "  }\n",
    "  \n",
    "  if (includePlots == FALSE & includecountMat == TRUE){\n",
    "    return(list(rse = rse_gene, vsdoutput = vsd, countMatoutput = countMat))\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7a37111",
   "metadata": {},
   "source": [
    "The next step is to create TCGA Random Summarized Experiment objects. One has only protein coding genes, another has only lincRNA genes, and another only has miRNA genes. Count matrices with variance-stabilization transform applied are also created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3736fdfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "vsdcreatorTumorOutput <- vsdcreator(rseobject = rse_geneTumorDataOriginal, genetype = \"protein_coding\", includePlots = TRUE, includecountMat = TRUE, isTumor = TRUE)\n",
    "\n",
    "rse_geneTumorData <- vsdcreatorTumorOutput[[1]]\n",
    "vsdTumorData <- vsdcreatorTumorOutput[[2]]\n",
    "\n",
    "vsdcreatorTumorOutputlincRNA <- vsdcreator(rseobject = rse_geneTumorDataOriginal, genetype = \"lincRNA\", includePlots = TRUE, includecountMat = TRUE, isTumor = TRUE)\n",
    "\n",
    "rse_geneTumorDatalincRNA <- vsdcreatorTumorOutputlincRNA[[1]]\n",
    "vsdTumorDatalincRNA <- vsdcreatorTumorOutputlincRNA[[2]]\n",
    "\n",
    "vsdcreatorTumorOutputmiRNA <- vsdcreator(rseobject = rse_geneTumorDataOriginal, genetype = \"miRNA\", includePlots = TRUE, includecountMat = TRUE, isTumor = TRUE)\n",
    "\n",
    "rse_geneTumorDatamiRNA <- vsdcreatorTumorOutputmiRNA[[1]]\n",
    "vsdTumorDatamiRNA <- vsdcreatorTumorOutputmiRNA[[2]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f35d25e",
   "metadata": {},
   "source": [
    "The next step is to create GTEx Random Summarized Experiment objects. One has only protein coding genes, another has only lincRNA genes, and another only has miRNA genes. Count matrices with variance-stabilization transform applied are also created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a03c1211",
   "metadata": {},
   "outputs": [],
   "source": [
    "vsdcreatorNormalOutput <- vsdcreator(rseobject = rse_geneNormalDataOriginal, genetype = \"protein_coding\", includePlots = TRUE, includecountMat = TRUE, isTumor = FALSE)\n",
    "\n",
    "rse_geneNormalData <- vsdcreatorNormalOutput[[1]]\n",
    "vsdNormalData <- vsdcreatorNormalOutput[[2]]\n",
    "\n",
    "vsdcreatorNormalOutputlincRNA <- vsdcreator(rseobject = rse_geneNormalDataOriginal, genetype = \"lincRNA\", includePlots = TRUE, includecountMat = TRUE, isTumor = FALSE)\n",
    "\n",
    "rse_geneNormalDatalincRNA <- vsdcreatorNormalOutputlincRNA[[1]]\n",
    "vsdNormalDatalincRNA <- vsdcreatorNormalOutputlincRNA[[2]]\n",
    "\n",
    "vsdcreatorNormalOutputmiRNA <- vsdcreator(rseobject = rse_geneNormalDataOriginal, genetype = \"miRNA\", includePlots = TRUE, includecountMat = TRUE, isTumor = FALSE)\n",
    "\n",
    "rse_geneNormalDatamiRNA <- vsdcreatorNormalOutputmiRNA[[1]]\n",
    "vsdNormalDatamiRNA <- vsdcreatorNormalOutputmiRNA[[2]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2cceb79",
   "metadata": {},
   "source": [
    "The functions \"tumorsexsubsetter\" and \"normalsexsubsetter\" take a Random Summarized Experiment object with TCGA data and a Random Summarized Experiment object with GTEx data respectively and subsets the samples by sex (\"M\" stands for male and \"F\" stands for female)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bab4e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "tumorsexsubsetter <- function(rseobject, sex){\n",
    "  if (sex == \"M\"){\n",
    "    tcgaidentifier = \"male\"\n",
    "  }\n",
    "  if (sex == \"F\"){\n",
    "    tcgaidentifier = \"female\"\n",
    "  }\n",
    "  \n",
    "  rse_gene <- rseobject\n",
    "  colDataRse <- colData(rseobject)\n",
    "  tumorIds = c()\n",
    "  patientSymbol = c()\n",
    "  \n",
    "  for(i in 1:length(colDataRse@listData$tcga.gdc_cases.demographic.gender)){\n",
    "    if((colDataRse@listData$tcga.gdc_cases.demographic.gender[i] %in% c(tcgaidentifier)) && (colDataRse@listData$tcga.gdc_cases.project.name[i] == \"Pancreatic Adenocarcinoma\")){\n",
    "      tumorIds = c(tumorIds,i)\n",
    "      patientSymbol = c(patientSymbol, colDataRse@listData$tcga.gdc_cases.samples.portions.analytes.aliquots.submitter_id[i])\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  rse_gene=rse_gene[,tumorIds]\n",
    "  colnames(rse_gene) = patientSymbol\n",
    "  \n",
    "  return(list(rse = rse_gene, Ids = tumorIds))\n",
    "  \n",
    "}\n",
    "\n",
    "normalsexsubsetter <- function(rseobject, sex){\n",
    "  if (sex == \"M\"){\n",
    "    gtexidentifier = 1\n",
    "  }\n",
    "  if (sex == \"F\"){\n",
    "    gtexidentifier = 2\n",
    "  }\n",
    "  \n",
    "  rse_gene <- rseobject\n",
    "  colDataRse <- colData(rseobject)\n",
    "  normalIds = c()\n",
    "  patientSymbol = c()\n",
    "  \n",
    "  for(i in 1:length(colDataRse@listData$gtex.sex)){\n",
    "    if((colDataRse@listData$gtex.sex[i] == gtexidentifier)){\n",
    "      normalIds = c(normalIds,i)     \n",
    "      patientSymbol = c(patientSymbol, colDataRse@listData$subjid[i])\n",
    "    }\n",
    "  }\n",
    "  rse_gene=rse_gene[,normalIds]\n",
    "  colnames(rse_gene) = paste0(colnames(rse_gene), colnamesidentifier)\n",
    "  \n",
    "  return(list(rse = rse_gene, Ids = normalIds))\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ca8985a",
   "metadata": {},
   "source": [
    "The function \"sexsubsettervsd\" subsets a count matrix with Variance Stabilization Transform applied so that it only contains the samples in the Random Summarized Experiment Object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d05f3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sexsubsettervsd <- function(rse_gene, vsd){\n",
    "  list <- rse_gene@colData@rownames\n",
    "  for (i in 1:length(list)){\n",
    "    list[i] <- substr(list[i], 1, nchar(list[i]))\n",
    "  }\n",
    "  vsdoutput <- vsd[,list]\n",
    "\n",
    "  return(vsdoutput)\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7c79c5f",
   "metadata": {},
   "source": [
    "In the next step we create the objects \"vsdMaleTumor,\" \"rse_geneMaleTumormiRNA,\" \"rse_geneMaleTumorlincRNA,\" \"vsdFeMaleTumor,\" \"rse_geneFeMaleTumormiRNA,\" and \"rse_geneFeMaleTumorlincRNA.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "667dd2fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein-coding genes\n",
    "tumorsexsubsetterTumorMaleOutput <-tumorsexsubsetter(rse_geneTumorData, \"M\")\n",
    "\n",
    "rse_geneMaleTumor <- tumorsexsubsetterTumorMaleOutput[[1]]\n",
    "\n",
    "vsdMaleTumor <- sexsubsettervsd(rse_geneMaleTumor, vsdTumorData)\n",
    "\n",
    "tumorsexsubsetterTumorFeMaleOutput <-tumorsexsubsetter(rse_geneTumorData, \"F\")\n",
    "\n",
    "rse_geneFeMaleTumor <- tumorsexsubsetterTumorFeMaleOutput[[1]]\n",
    "\n",
    "vsdFeMaleTumor <- sexsubsettervsd(rse_geneFeMaleTumor, vsdTumorData)\n",
    "\n",
    "# miRNA genes\n",
    "\n",
    "tumorsexsubsetterTumorMaleOutputmiRNA <-tumorsexsubsetter(rse_geneTumorDatamiRNA, \"M\")\n",
    "\n",
    "rse_geneMaleTumormiRNA <- tumorsexsubsetterTumorMaleOutputmiRNA[[1]]\n",
    "\n",
    "vsdMaleTumormiRNA <- sexsubsettervsd(rse_geneMaleTumormiRNA, vsdTumorDatamiRNA)\n",
    "\n",
    "tumorsexsubsetterTumorFeMaleOutputmiRNA <-tumorsexsubsetter(rse_geneTumorDatamiRNA, \"F\")\n",
    "\n",
    "rse_geneFeMaleTumormiRNA <- tumorsexsubsetterTumorFeMaleOutputmiRNA[[1]]\n",
    "\n",
    "vsdFeMaleTumormiRNA <- sexsubsettervsd(rse_geneFeMaleTumormiRNA, vsdTumorDatamiRNA)\n",
    "\n",
    "# lincRNA genes\n",
    "\n",
    "tumorsexsubsetterTumorMaleOutputlincRNA <-tumorsexsubsetter(rse_geneTumorDatalincRNA, \"M\")\n",
    "\n",
    "rse_geneMaleTumorlincRNA <- tumorsexsubsetterTumorMaleOutputlincRNA[[1]]\n",
    "\n",
    "vsdMaleTumorlincRNA <- sexsubsettervsd(rse_geneMaleTumorlincRNA, vsdTumorDatalincRNA)\n",
    "\n",
    "tumorsexsubsetterTumorFeMaleOutputlincRNA <-tumorsexsubsetter(rse_geneTumorDatalincRNA, \"F\")\n",
    "\n",
    "rse_geneFeMaleTumorlincRNA <- tumorsexsubsetterTumorFeMaleOutputlincRNA[[1]]\n",
    "\n",
    "vsdFeMaleTumorlincRNA <- sexsubsettervsd(rse_geneFeMaleTumorlincRNA, vsdTumorDatalincRNA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd790e4d",
   "metadata": {},
   "source": [
    "In the next step we create the objects \"vsdMaleNormal,\" \"rse_geneMaleNormalmiRNA,\" \"rse_geneMaleNormallincRNA,\" \"vsdFeMaleNormal,\" \"rse_geneFeMaleNormalmiRNA,\" and \"rse_geneFeMaleNormallincRNA.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bce0740b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein-coding genes\n",
    "\n",
    "normalsexsubsetterNormalMaleOutput <- normalsexsubsetter(rse_geneNormalData, \"M\")\n",
    "\n",
    "rse_geneMaleNormal <- normalsexsubsetterNormalMaleOutput[[1]]\n",
    "MaleNormalIds <- normalsexsubsetterNormalMaleOutput[[2]]\n",
    "\n",
    "vsdMaleNormal <- sexsubsettervsd(rse_geneMaleNormal, vsdNormalData)\n",
    "\n",
    "normalsexsubsetterNormalFeMaleOutput <-normalsexsubsetter(rse_geneNormalData, \"F\")\n",
    "\n",
    "rse_geneFeMaleNormal <- normalsexsubsetterNormalFeMaleOutput[[1]]\n",
    "FeMaleNormalIds <- normalsexsubsetterNormalFeMaleOutput[[2]]\n",
    "\n",
    "vsdFeMaleNormal <- sexsubsettervsd(rse_geneFeMaleNormal, vsdNormalData)\n",
    "\n",
    "# miRNA genes\n",
    "\n",
    "normalsexsubsetterNormalMaleOutputmiRNA <-normalsexsubsetter(rse_geneNormalDatamiRNA, \"M\")\n",
    "\n",
    "rse_geneMaleNormalmiRNA <- normalsexsubsetterNormalMaleOutputmiRNA[[1]]\n",
    "\n",
    "vsdMaleNormalmiRNA <- sexsubsettervsd(rse_geneMaleNormalmiRNA, vsdNormalDatamiRNA)\n",
    "\n",
    "normalsexsubsetterNormalFeMaleOutputmiRNA <-normalsexsubsetter(rse_geneNormalDatamiRNA, \"F\")\n",
    "\n",
    "rse_geneFeMaleNormalmiRNA <- normalsexsubsetterNormalFeMaleOutputmiRNA[[1]]\n",
    "\n",
    "vsdFeMaleNormalmiRNA <- sexsubsettervsd(rse_geneFeMaleNormalmiRNA, vsdNormalDatamiRNA)\n",
    "\n",
    "# lincRNA genes\n",
    "\n",
    "normalsexsubsetterNormalMaleOutputlincRNA <-normalsexsubsetter(rse_geneNormalDatalincRNA, \"M\")\n",
    "\n",
    "rse_geneMaleNormallincRNA <- normalsexsubsetterNormalMaleOutputlincRNA[[1]]\n",
    "\n",
    "vsdMaleNormallincRNA <- sexsubsettervsd(rse_geneMaleNormallincRNA, vsdNormalDatalincRNA)\n",
    "\n",
    "normalsexsubsetterNormalFeMaleOutputlincRNA <-normalsexsubsetter(rse_geneNormalDatalincRNA, \"F\")\n",
    "\n",
    "rse_geneFeMaleNormallincRNA <- normalsexsubsetterNormalFeMaleOutputlincRNA[[1]]\n",
    "\n",
    "vsdFeMaleNormallincRNA <- sexsubsettervsd(rse_geneFeMaleNormallincRNA, vsdNormalDatalincRNA)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca845114",
   "metadata": {},
   "source": [
    "In the next step we create a visnetwork graph displaying the results of a differential gene regulatory network analysis between gene regulatory networks created using TCGA samples and GTEx Samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "809ae480",
   "metadata": {},
   "outputs": [],
   "source": [
    "NormalAllMatrix <- as.data.frame(t(rbind(vsdcreator(rseobject = rse_geneNormalDataOriginal, genetype = \"protein_coding\", includePlots = FALSE, includecountMat = FALSE, isTumor = FALSE)[[2]], vsdcreator(rseobject = rse_geneNormalDataOriginal, genetype = \"lincRNA\", includePlots = FALSE, includecountMat = FALSE, isTumor = FALSE)[[2]], vsdcreator(rseobject = rse_geneNormalDataOriginal, genetype = \"miRNA\", includePlots = FALSE, includecountMat = FALSE, isTumor = FALSE)[[2]])))\n",
    "\n",
    "TumorAllMatrix <- as.data.frame(t(rbind(vsdcreator(rseobject = rse_geneTumorDataOriginal, genetype = \"protein_coding\", includePlots = FALSE, includecountMat = FALSE, isTumor = TRUE)[[2]], vsdcreator(rseobject = rse_geneTumorDataOriginal, genetype = \"lincRNA\", includePlots = FALSE, includecountMat = FALSE, isTumor = TRUE)[[2]], vsdcreator(rseobject = rse_geneTumorDataOriginal, genetype = \"miRNA\", includePlots = FALSE, includecountMat = FALSE, isTumor = TRUE)[[2]])))\n",
    "\n",
    "ListNATAComp <- intersect(colnames(NormalAllMatrix), colnames(TumorAllMatrix))\n",
    "\n",
    "CorNormalAllMatrix <- as.data.frame(cor(NormalAllMatrix[,ListNATAComp]))\n",
    "CorTumorAllMatrix <- as.data.frame(cor(TumorAllMatrix[,ListNATAComp]))\n",
    "\n",
    "\n",
    "CorNormalAllMatrix <- CorNormalAllMatrix %>% select(as.vector(intersect(ListNATAComp, rownames(vsdNormalData))))\n",
    "CorNormalAllMatrix <- CorNormalAllMatrix %>% select(-as.vector(intersect(intersect(ListNATAComp, rownames(vsdNormalData)), human_tf)))\n",
    "CorNormalAllMatrix <- as.data.frame(subset(CorNormalAllMatrix, !(rownames(CorNormalAllMatrix) %in% colnames(CorNormalAllMatrix))))\n",
    "\n",
    "CorTumorAllMatrix <- CorTumorAllMatrix %>% select(as.vector(intersect(ListNATAComp, rownames(vsdTumorData))))\n",
    "CorTumorAllMatrix <- CorTumorAllMatrix %>% select(-as.vector(intersect(intersect(ListNATAComp, rownames(vsdTumorData)), human_tf)))\n",
    "CorTumorAllMatrix <- as.data.frame(subset(CorTumorAllMatrix, !(rownames(CorTumorAllMatrix) %in% colnames(CorTumorAllMatrix))))\n",
    "\n",
    "AllComparisonMatrix <- CorTumorAllMatrix - CorNormalAllMatrix\n",
    "AllComparisonMatrixrownames = rownames(AllComparisonMatrix)\n",
    "\n",
    "num_edges <- 100\n",
    "\n",
    "edges = matrix(0L, num_edges, 3)\n",
    "colnames(edges) = c(\"from\",\"to\",\"value\")\n",
    "edges = as.data.frame(edges)\n",
    "\n",
    "sort_mat = order(as.matrix(abs(AllComparisonMatrix)), decreasing = TRUE)\n",
    "edges$value  = as.matrix(AllComparisonMatrix)[sort_mat[1:num_edges]]\n",
    "\n",
    "pcgeneIdsTop = (sort_mat[1:num_edges] %/% dim(AllComparisonMatrix)[1]) + 1\n",
    "rnaIdsTop = sort_mat[1:num_edges] %% dim(AllComparisonMatrix)[1]\n",
    "nrnas = dim(AllComparisonMatrix)[1]\n",
    "rnaIdsTop[rnaIdsTop == 0] = nrnas\n",
    "\n",
    "edges$to = colnames(AllComparisonMatrix)[pcgeneIdsTop]\n",
    "edges$from = rownames(AllComparisonMatrix)[rnaIdsTop]\n",
    "edges$arrows = \"to\"\n",
    "edges$color = ifelse(edges$value > 0, \"green\", \"red\")\n",
    "edges$value = abs(edges$value)\n",
    "\n",
    "nodes = data.frame(id = unique(as.vector(as.matrix(edges[,c(1,2)]))), label=unique(as.vector(as.matrix(edges[,c(1,2)]))))\n",
    "#nodes$group = ifelse(nodes$id %in% edges$from, \"miRNA/lincRNA\", \"Protein-coding Gene\")\n",
    "idmirna <- c()\n",
    "idlncrna <- c()\n",
    "idtfs <- c()\n",
    "mirnavector <- as.vector(union(rownames(vsdNormalDatamiRNA),rownames(vsdTumorDatamiRNA)))\n",
    "lncrnavector <- as.vector(union(rownames(vsdNormalDatalincRNA),rownames(vsdTumorDatalincRNA)))\n",
    "\n",
    "for(i in 1:dim(nodes)[1]){\n",
    "  if(nodes[i,1] %in% human_tf){\n",
    "    idtfs = c(idtfs, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% mirnavector){\n",
    "    idmirna = c(idmirna, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% lncrnavector){\n",
    "    idlncrna = c(idlncrna, i)\n",
    "  }\n",
    "}\n",
    "nodes$group = 'Protein-coding Gene'\n",
    "nodes$group[idmirna] = 'miRNA'\n",
    "nodes$group[idlncrna] = 'lncRNA'\n",
    "nodes$group[idtfs] = 'tfs'\n",
    "\n",
    "net <- visNetwork(nodes, edges, width = \"100%\")\n",
    "net <- visGroups(net, groupname = \"tfs\", shape = \"triangle\",\n",
    "                 color = list(background = \"green\", border=\"green\"))\n",
    "net <- visGroups(net, groupname = \"lncRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"pink\", border=\"pink\"))\n",
    "net <- visGroups(net, groupname = \"miRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"orange\", border=\"orange\"))\n",
    "net <- visGroups(net, groupname = \"Protein-coding Gene\", shape = \"dot\",       \n",
    "                 color = list(background = \"darkblue\", border=\"black\"))\n",
    "visLegend(net, main=\"Legend\", position=\"right\", ncol=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e06afc65",
   "metadata": {},
   "source": [
    "In the next step we create a visnetwork graph displaying the results of a differential gene regulatory network analysis between gene regulatory networks created using male TCGA samples and male GTEx Samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c4310b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "NormalMaleMatrix <- as.data.frame(t(rbind(vsdMaleNormal, vsdMaleNormalmiRNA, vsdMaleNormallincRNA)))\n",
    "\n",
    "TumorMaleMatrix <- as.data.frame(t(rbind(vsdMaleTumor, vsdMaleTumormiRNA, vsdMaleTumorlincRNA)))\n",
    "\n",
    "ListNMTMComp <- intersect(colnames(NormalMaleMatrix), colnames(TumorMaleMatrix))\n",
    "\n",
    "CorNormalMaleMatrix <- as.data.frame(cor(NormalMaleMatrix[,ListNMTMComp]))\n",
    "CorTumorMaleMatrix <- as.data.frame(cor(TumorMaleMatrix[,ListNMTMComp]))\n",
    "\n",
    "\n",
    "CorNormalMaleMatrix <- CorNormalMaleMatrix %>% select(as.vector(intersect(ListNMTMComp, rownames(vsdMaleNormal))))\n",
    "CorNormalMaleMatrix <- CorNormalMaleMatrix %>% select(-as.vector(intersect(intersect(ListNMTMComp, rownames(vsdMaleNormal)), human_tf)))\n",
    "CorNormalMaleMatrix <- as.data.frame(subset(CorNormalMaleMatrix, !(rownames(CorNormalMaleMatrix) %in% colnames(CorNormalMaleMatrix))))\n",
    "\n",
    "CorTumorMaleMatrix <- CorTumorMaleMatrix %>% select(as.vector(intersect(ListNMTMComp, rownames(vsdMaleTumor))))\n",
    "CorTumorMaleMatrix <- CorTumorMaleMatrix %>% select(-as.vector(intersect(intersect(ListNMTMComp, rownames(vsdMaleTumor)), human_tf)))\n",
    "CorTumorMaleMatrix <- as.data.frame(subset(CorTumorMaleMatrix, !(rownames(CorTumorMaleMatrix) %in% colnames(CorTumorMaleMatrix))))\n",
    "\n",
    "MaleComparisonMatrix <- CorTumorMaleMatrix - CorNormalMaleMatrix\n",
    "MaleComparisonMatrixrownames = rownames(MaleComparisonMatrix)\n",
    "\n",
    "num_edges <- 100\n",
    "\n",
    "edges = matrix(0L, num_edges, 3)\n",
    "colnames(edges) = c(\"from\",\"to\",\"value\")\n",
    "edges = as.data.frame(edges)\n",
    "\n",
    "sort_mat = order(as.matrix(abs(MaleComparisonMatrix)), decreasing = TRUE)\n",
    "edges$value  = as.matrix(MaleComparisonMatrix)[sort_mat[1:num_edges]]\n",
    "\n",
    "pcgeneIdsTop = (sort_mat[1:num_edges] %/% dim(MaleComparisonMatrix)[1]) + 1\n",
    "rnaIdsTop = sort_mat[1:num_edges] %% dim(MaleComparisonMatrix)[1]\n",
    "nrnas = dim(MaleComparisonMatrix)[1]\n",
    "rnaIdsTop[rnaIdsTop == 0] = nrnas\n",
    "\n",
    "edges$to = colnames(MaleComparisonMatrix)[pcgeneIdsTop]\n",
    "edges$from = rownames(MaleComparisonMatrix)[rnaIdsTop]\n",
    "edges$arrows = \"to\"\n",
    "edges$color = ifelse(edges$value > 0, \"green\", \"red\")\n",
    "edges$value = abs(edges$value)\n",
    "\n",
    "nodes = data.frame(id = unique(as.vector(as.matrix(edges[,c(1,2)]))), label=unique(as.vector(as.matrix(edges[,c(1,2)]))))\n",
    "#nodes$group = ifelse(nodes$id %in% edges$from, \"miRNA/lincRNA\", \"Protein-coding Gene\")\n",
    "idmirna <- c()\n",
    "idlncrna <- c()\n",
    "idtfs <- c()\n",
    "mirnavector <- as.vector(union(rownames(vsdMaleNormalmiRNA),rownames(vsdMaleTumormiRNA)))\n",
    "lncrnavector <- as.vector(union(rownames(vsdMaleNormallincRNA),rownames(vsdMaleTumorlincRNA)))\n",
    "\n",
    "for(i in 1:dim(nodes)[1]){\n",
    "  if(nodes[i,1] %in% human_tf){\n",
    "    idtfs = c(idtfs, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% mirnavector){\n",
    "    idmirna = c(idmirna, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% lncrnavector){\n",
    "    idlncrna = c(idlncrna, i)\n",
    "  }\n",
    "}\n",
    "nodes$group = 'Protein-coding Gene'\n",
    "nodes$group[idmirna] = 'miRNA'\n",
    "nodes$group[idlncrna] = 'lncRNA'\n",
    "nodes$group[idtfs] = 'tfs'\n",
    "\n",
    "net <- visNetwork(nodes, edges, width = \"100%\")\n",
    "net <- visGroups(net, groupname = \"tfs\", shape = \"triangle\",\n",
    "                 color = list(background = \"green\", border=\"green\"))\n",
    "net <- visGroups(net, groupname = \"lncRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"pink\", border=\"pink\"))\n",
    "net <- visGroups(net, groupname = \"miRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"orange\", border=\"orange\"))\n",
    "net <- visGroups(net, groupname = \"Protein-coding Gene\", shape = \"dot\",       \n",
    "                 color = list(background = \"darkblue\", border=\"black\"))\n",
    "visLegend(net, main=\"Legend\", position=\"right\", ncol=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71ae6b22",
   "metadata": {},
   "source": [
    "In the next step we create a visnetwork graph displaying the results of a differential gene regulatory network analysis between gene regulatory networks created using female TCGA samples and female GTEx Samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8e16eb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "NormalFeMaleMatrix <- as.data.frame(t(rbind(vsdFeMaleNormal, vsdFeMaleNormalmiRNA, vsdFeMaleNormallincRNA)))\n",
    "\n",
    "TumorFeMaleMatrix <- as.data.frame(t(rbind(vsdFeMaleTumor, vsdFeMaleTumormiRNA, vsdFeMaleTumorlincRNA)))\n",
    "\n",
    "ListNFTFComp <- intersect(colnames(NormalFeMaleMatrix), colnames(TumorFeMaleMatrix))\n",
    "\n",
    "CorNormalFeMaleMatrix <- as.data.frame(cor(NormalFeMaleMatrix[,ListNFTFComp]))\n",
    "CorTumorFeMaleMatrix <- as.data.frame(cor(TumorFeMaleMatrix[,ListNFTFComp]))\n",
    "\n",
    "CorNormalFeMaleMatrix <- CorNormalFeMaleMatrix %>% select(as.vector(intersect(ListNFTFComp, rownames(vsdFeMaleNormal))))\n",
    "CorNormalFeMaleMatrix <- CorNormalFeMaleMatrix %>% select(-as.vector(intersect(intersect(ListNFTFComp, rownames(vsdFeMaleNormal)), human_tf)))\n",
    "CorNormalFeMaleMatrix <- as.data.frame(subset(CorNormalFeMaleMatrix, !(rownames(CorNormalFeMaleMatrix) %in% colnames(CorNormalFeMaleMatrix))))\n",
    "\n",
    "CorTumorFeMaleMatrix <- CorTumorFeMaleMatrix %>% select(as.vector(intersect(ListNFTFComp, rownames(vsdFeMaleTumor))))\n",
    "CorTumorFeMaleMatrix <- CorTumorFeMaleMatrix %>% select(-as.vector(intersect(intersect(ListNFTFComp, rownames(vsdFeMaleTumor)), human_tf)))\n",
    "CorTumorFeMaleMatrix <- as.data.frame(subset(CorTumorFeMaleMatrix, !(rownames(CorTumorFeMaleMatrix) %in% colnames(CorTumorFeMaleMatrix))))\n",
    "\n",
    "FeMaleComparisonMatrix <- CorTumorFeMaleMatrix - CorNormalFeMaleMatrix\n",
    "FeMaleComparisonMatrixrownames = rownames(FeMaleComparisonMatrix)\n",
    "\n",
    "num_edges <- 100\n",
    "\n",
    "edges = matrix(0L, num_edges, 3)\n",
    "colnames(edges) = c(\"from\",\"to\",\"value\")\n",
    "edges = as.data.frame(edges)\n",
    "\n",
    "sort_mat = order(as.matrix(abs(FeMaleComparisonMatrix)), decreasing = TRUE)\n",
    "edges$value  = as.matrix(FeMaleComparisonMatrix)[sort_mat[1:num_edges]]\n",
    "\n",
    "pcgeneIdsTop = (sort_mat[1:num_edges] %/% dim(FeMaleComparisonMatrix)[1]) + 1\n",
    "rnaIdsTop = sort_mat[1:num_edges] %% dim(FeMaleComparisonMatrix)[1]\n",
    "nrnas = dim(FeMaleComparisonMatrix)[1]\n",
    "rnaIdsTop[rnaIdsTop == 0] = nrnas\n",
    "\n",
    "edges$to = colnames(FeMaleComparisonMatrix)[pcgeneIdsTop]\n",
    "edges$from = rownames(FeMaleComparisonMatrix)[rnaIdsTop]\n",
    "edges$arrows = \"to\"\n",
    "edges$color = ifelse(edges$value > 0, \"green\", \"red\")\n",
    "edges$value = abs(edges$value)\n",
    "\n",
    "nodes = data.frame(id = unique(as.vector(as.matrix(edges[,c(1,2)]))), label=unique(as.vector(as.matrix(edges[,c(1,2)]))))\n",
    "#nodes$group = ifelse(nodes$id %in% edges$from, \"miRNA/lincRNA\", \"Protein-coding Gene\")\n",
    "idmirna <- c()\n",
    "idlncrna <- c()\n",
    "idtfs <- c()\n",
    "mirnavector <- as.vector(union(rownames(vsdFeMaleNormalmiRNA),rownames(vsdFeMaleTumormiRNA)))\n",
    "lncrnavector <- as.vector(union(rownames(vsdFeMaleNormallincRNA),rownames(vsdFeMaleTumorlincRNA)))\n",
    "\n",
    "for(i in 1:dim(nodes)[1]){\n",
    "  if(nodes[i,1] %in% human_tf){\n",
    "    idtfs = c(idtfs, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% mirnavector){\n",
    "    idmirna = c(idmirna, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% lncrnavector){\n",
    "    idlncrna = c(idlncrna, i)\n",
    "  }\n",
    "}\n",
    "nodes$group = 'Protein-coding Gene'\n",
    "nodes$group[idmirna] = 'miRNA'\n",
    "nodes$group[idlncrna] = 'lncRNA'\n",
    "nodes$group[idtfs] = 'tfs'\n",
    "\n",
    "net <- visNetwork(nodes, edges, width = \"100%\")\n",
    "net <- visGroups(net, groupname = \"tfs\", shape = \"triangle\",\n",
    "                 color = list(background = \"green\", border=\"green\"))\n",
    "net <- visGroups(net, groupname = \"lncRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"pink\", border=\"pink\"))\n",
    "net <- visGroups(net, groupname = \"miRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"orange\", border=\"orange\"))\n",
    "net <- visGroups(net, groupname = \"Protein-coding Gene\", shape = \"dot\",       \n",
    "                 color = list(background = \"darkblue\", border=\"black\"))\n",
    "visLegend(net, main=\"Legend\", position=\"right\", ncol=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64ada952",
   "metadata": {},
   "source": [
    "In the next step we create a visnetwork graph displaying the results of a differential gene regulatory network analysis between a gene regulatory network created using Male GTEx samples and Female GTEx samples. Note that X and Y chromosome genes have been removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd42236a",
   "metadata": {},
   "outputs": [],
   "source": [
    "NormalMaleMatrix <- as.data.frame(t(rbind(vsdMaleNormal, vsdMaleNormalmiRNA, vsdMaleNormallincRNA)))\n",
    "\n",
    "NormalFeMaleMatrix <- as.data.frame(t(rbind(vsdFeMaleNormal, vsdFeMaleNormalmiRNA, vsdFeMaleNormallincRNA)))\n",
    "\n",
    "ListNMNFComp <- intersect(colnames(NormalMaleMatrix), colnames(NormalFeMaleMatrix))\n",
    "ListNMNFComp <- ListTMTFComp[ListTMTFComp %in% Xgene]\n",
    "ListNMNFComp <- ListTMTFComp[ListTMTFComp %in% Ygene]\n",
    "\n",
    "CorNormalMaleMatrix <- as.data.frame(cor(NormalMaleMatrix[,ListNMNFComp]))\n",
    "CorNormalFeMaleMatrix <- as.data.frame(cor(NormalFeMaleMatrix[,ListNMNFComp]))\n",
    "\n",
    "CorNormalMaleMatrix <- CorNormalMaleMatrix %>% select(as.vector(intersect(ListNMNFComp, rownames(vsdMaleNormal))))\n",
    "CorNormalMaleMatrix <- CorNormalMaleMatrix %>% select(-as.vector(intersect(intersect(ListNMNFComp, rownames(vsdMaleNormal)), human_tf)))\n",
    "CorNormalMaleMatrix <- as.data.frame(subset(CorNormalMaleMatrix, !(rownames(CorNormalMaleMatrix) %in% colnames(CorNormalMaleMatrix))))\n",
    "\n",
    "CorNormalFeMaleMatrix <- CorNormalFeMaleMatrix %>% select(as.vector(intersect(ListNMNFComp, rownames(vsdFeMaleNormal))))\n",
    "CorNormalFeMaleMatrix <- CorNormalFeMaleMatrix %>% select(-as.vector(intersect(intersect(ListNMNFComp, rownames(vsdFeMaleNormal)), human_tf)))\n",
    "CorNormalFeMaleMatrix <- as.data.frame(subset(CorNormalFeMaleMatrix, !(rownames(CorNormalFeMaleMatrix) %in% colnames(CorNormalFeMaleMatrix))))\n",
    "\n",
    "\n",
    "NormalComparisonMatrix <- CorNormalMaleMatrix - CorNormalFeMaleMatrix\n",
    "NormalComparisonMatrixrownames = rownames(NormalComparisonMatrix)\n",
    "\n",
    "num_edges <- 100\n",
    "\n",
    "edges = matrix(0L, num_edges, 3)\n",
    "colnames(edges) = c(\"from\",\"to\",\"value\")\n",
    "edges = as.data.frame(edges)\n",
    "\n",
    "sort_mat = order(as.matrix(abs(NormalComparisonMatrix)), decreasing = TRUE)\n",
    "edges$value  = as.matrix(NormalComparisonMatrix)[sort_mat[1:num_edges]]\n",
    "\n",
    "pcgeneIdsTop = (sort_mat[1:num_edges] %/% dim(NormalComparisonMatrix)[1]) + 1\n",
    "rnaIdsTop = sort_mat[1:num_edges] %% dim(NormalComparisonMatrix)[1]\n",
    "nrnas = dim(NormalComparisonMatrix)[1]\n",
    "rnaIdsTop[rnaIdsTop == 0] = nrnas\n",
    "\n",
    "edges$to = colnames(NormalComparisonMatrix)[pcgeneIdsTop]\n",
    "edges$from = rownames(NormalComparisonMatrix)[rnaIdsTop]\n",
    "edges$arrows = \"to\"\n",
    "edges$color = ifelse(edges$value > 0, \"green\", \"red\")\n",
    "edges$value = abs(edges$value)\n",
    "\n",
    "nodes = data.frame(id = unique(as.vector(as.matrix(edges[,c(1,2)]))), label=unique(as.vector(as.matrix(edges[,c(1,2)]))))\n",
    "#nodes$group = ifelse(nodes$id %in% edges$from, \"miRNA/lincRNA\", \"Protein-coding Gene\")\n",
    "idmirna <- c()\n",
    "idlncrna <- c()\n",
    "idtfs <- c()\n",
    "mirnavector <- as.vector(rownames(vsdNormalDatamiRNA))\n",
    "lncrnavector <- as.vector(rownames(vsdNormalDatamiRNA))\n",
    "\n",
    "for(i in 1:dim(nodes)[1]){\n",
    "  if(nodes[i,1] %in% human_tf){\n",
    "    idtfs = c(idtfs, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% mirnavector){\n",
    "    idmirna = c(idmirna, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% lncrnavector){\n",
    "    idlncrna = c(idlncrna, i)\n",
    "  }\n",
    "}\n",
    "nodes$group = 'Protein-coding Gene'\n",
    "nodes$group[idmirna] = 'miRNA'\n",
    "nodes$group[idlncrna] = 'lncRNA'\n",
    "nodes$group[idtfs] = 'tfs'\n",
    "\n",
    "net <- visNetwork(nodes, edges, width = \"100%\")\n",
    "net <- visGroups(net, groupname = \"tfs\", shape = \"triangle\",\n",
    "                 color = list(background = \"green\", border=\"green\"))\n",
    "net <- visGroups(net, groupname = \"lncRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"pink\", border=\"pink\"))\n",
    "net <- visGroups(net, groupname = \"miRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"orange\", border=\"orange\"))\n",
    "net <- visGroups(net, groupname = \"Protein-coding Gene\", shape = \"dot\",       \n",
    "                 color = list(background = \"darkblue\", border=\"black\"))\n",
    "visLegend(net, main=\"Legend\", position=\"right\", ncol=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7a982f9",
   "metadata": {},
   "source": [
    "In the next step we create a visnetwork graph displaying the results of a differential gene regulatory network analysis between a gene regulatory network created using Male TCGA samples and Female TCGA samples. Note that X and Y chromosome genes have been removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c25d2d8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "TumorMaleMatrix <- as.data.frame(t(rbind(vsdMaleTumor, vsdMaleTumormiRNA, vsdMaleTumorlincRNA)))\n",
    "\n",
    "TumorFeMaleMatrix <- as.data.frame(t(rbind(vsdFeMaleTumor, vsdFeMaleTumormiRNA, vsdFeMaleTumorlincRNA)))\n",
    "\n",
    "ListTMTFComp <- intersect(colnames(TumorMaleMatrix), colnames(TumorFeMaleMatrix))\n",
    "ListTMTFComp <- ListTMTFComp[ListTMTFComp %in% Xgene]\n",
    "ListTMTFComp <- ListTMTFComp[ListTMTFComp %in% Ygene]\n",
    "\n",
    "CorTumorMaleMatrix <- as.data.frame(cor(TumorMaleMatrix[,ListTMTFComp]))\n",
    "CorTumorFeMaleMatrix <- as.data.frame(cor(TumorFeMaleMatrix[,ListTMTFComp]))\n",
    "\n",
    "CorTumorMaleMatrix <- CorTumorMaleMatrix %>% select(as.vector(intersect(ListTMTFComp, rownames(vsdMaleTumor))))\n",
    "CorTumorMaleMatrix <- CorTumorMaleMatrix %>% select(-as.vector(intersect(intersect(ListTMTFComp, rownames(vsdMaleTumor)), human_tf)))\n",
    "CorTumorMaleMatrix <- as.data.frame(subset(CorTumorMaleMatrix, !(rownames(CorTumorMaleMatrix) %in% colnames(CorTumorMaleMatrix))))\n",
    "\n",
    "CorTumorFeMaleMatrix <- CorTumorFeMaleMatrix %>% select(as.vector(intersect(ListTMTFComp, rownames(vsdFeMaleTumor))))\n",
    "CorTumorFeMaleMatrix <- CorTumorFeMaleMatrix %>% select(-as.vector(intersect(intersect(ListTMTFComp, rownames(vsdFeMaleTumor)), human_tf)))\n",
    "CorTumorFeMaleMatrix <- as.data.frame(subset(CorTumorFeMaleMatrix, !(rownames(CorTumorFeMaleMatrix) %in% colnames(CorTumorFeMaleMatrix))))\n",
    "\n",
    "\n",
    "TumorComparisonMatrix <- CorTumorMaleMatrix - CorTumorFeMaleMatrix\n",
    "TumorComparisonMatrixrownames = rownames(TumorComparisonMatrix)\n",
    "\n",
    "num_edges <- 100\n",
    "\n",
    "edges = matrix(0L, num_edges, 3)\n",
    "colnames(edges) = c(\"from\",\"to\",\"value\")\n",
    "edges = as.data.frame(edges)\n",
    "\n",
    "sort_mat = order(as.matrix(abs(TumorComparisonMatrix)), decreasing = TRUE)\n",
    "edges$value  = as.matrix(TumorComparisonMatrix)[sort_mat[1:num_edges]]\n",
    "\n",
    "pcgeneIdsTop = (sort_mat[1:num_edges] %/% dim(TumorComparisonMatrix)[1]) + 1\n",
    "rnaIdsTop = sort_mat[1:num_edges] %% dim(TumorComparisonMatrix)[1]\n",
    "nrnas = dim(TumorComparisonMatrix)[1]\n",
    "rnaIdsTop[rnaIdsTop == 0] = nrnas\n",
    "\n",
    "edges$to = colnames(TumorComparisonMatrix)[pcgeneIdsTop]\n",
    "edges$from = rownames(TumorComparisonMatrix)[rnaIdsTop]\n",
    "edges$arrows = \"to\"\n",
    "edges$color = ifelse(edges$value > 0, \"green\", \"red\")\n",
    "edges$value = abs(edges$value)\n",
    "\n",
    "nodes = data.frame(id = unique(as.vector(as.matrix(edges[,c(1,2)]))), label=unique(as.vector(as.matrix(edges[,c(1,2)]))))\n",
    "#nodes$group = ifelse(nodes$id %in% edges$from, \"miRNA/lincRNA\", \"Protein-coding Gene\")\n",
    "idmirna <- c()\n",
    "idlncrna <- c()\n",
    "idtfs <- c()\n",
    "mirnavector <- as.vector(rownames(vsdTumorDatamiRNA))\n",
    "lncrnavector <- as.vector(rownames(vsdTumorDatamiRNA))\n",
    "\n",
    "for(i in 1:dim(nodes)[1]){\n",
    "  if(nodes[i,1] %in% human_tf){\n",
    "    idtfs = c(idtfs, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% mirnavector){\n",
    "    idmirna = c(idmirna, i)\n",
    "  }\n",
    "  if(nodes[i,1] %in% lncrnavector){\n",
    "    idlncrna = c(idlncrna, i)\n",
    "  }\n",
    "}\n",
    "nodes$group = 'Protein-coding Gene'\n",
    "nodes$group[idmirna] = 'miRNA'\n",
    "nodes$group[idlncrna] = 'lncRNA'\n",
    "nodes$group[idtfs] = 'tfs'\n",
    "\n",
    "net <- visNetwork(nodes, edges, width = \"100%\")\n",
    "net <- visGroups(net, groupname = \"tfs\", shape = \"triangle\",\n",
    "                 color = list(background = \"green\", border=\"green\"))\n",
    "net <- visGroups(net, groupname = \"lncRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"pink\", border=\"pink\"))\n",
    "net <- visGroups(net, groupname = \"miRNA\", shape = \"triangle\",\n",
    "                 color = list(background = \"orange\", border=\"orange\"))\n",
    "net <- visGroups(net, groupname = \"Protein-coding Gene\", shape = \"dot\",       \n",
    "                 color = list(background = \"darkblue\", border=\"black\"))\n",
    "visLegend(net, main=\"Legend\", position=\"right\", ncol=1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
